bindings/go/ctf/filesystem_ctf.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	return c.writeFile(v1.ArtifactIndexFileName, bytes.NewReader(data))
+}
+
+// writeFile writes the given raw data to the given name in the CTF.
+// If the directory does not exist, it will be created.
+func (c *FileSystemCTF) writeFile(name string, raw io.Reader) (err error) {
+	if err := c.fs.MkdirAll(filepath.Dir(name), 0o755); err != nil {
+		return fmt.Errorf("unable to create directory: %w", err)
+	}
+	var file fs.File
+	if file, err = c.fs.OpenFile(name, os.O_CREATE|os.O_WRONLY, 0o644); err != nil {
+		return fmt.Errorf("unable to open artifact index: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, file.Close())
+	}()
+
+	writeable, ok := file.(io.Writer)
+	if !ok {
+		return fmt.Errorf("file %s is read only and cannot be saved", name)
+	}
+
+	buf := ioBufPool.Get().(*[]byte)
+	defer ioBufPool.Put(buf)
+	if _, err = io.CopyBuffer(writeable, raw, *buf); err != nil {
+		return fmt.Errorf("unable to write artifact index: %w", err)
+	}
+
+	return nil
+}
+
+// DeleteBlob deletes the blob with the given digest from the CTF by removing the file from BlobsDirectoryName.
+func (c *FileSystemCTF) DeleteBlob(digest string) (err error) {
+	if err = c.fs.Remove(filepath.Join(BlobsDirectoryName, ToBlobFileName(digest))); err != nil {
+		return fmt.Errorf("unable to delete blob: %w", err)
+	}
+
+	return nil
+}
+
+// GetBlob returns the blob with the given digest from the CTF by reading the file from BlobsDirectoryName.
+func (c *FileSystemCTF) GetBlob(digest string) (blob.ReadOnlyBlob, error) {
+	b := NewCASFileBlob(c.fs, filepath.Join(BlobsDirectoryName, ToBlobFileName(digest)))
+	b.SetPrecalculatedDigest(digest)
+	return b, nil
+}
+
+// ListBlobs returns a list of all blobs in the CTF by listing the files in BlobsDirectoryName.
+func (c *FileSystemCTF) ListBlobs() (digests []string, err error) {
+	dir, err := c.fs.ReadDir(BlobsDirectoryName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to list blobs: %w", err)
+	}
+
+	digests = make([]string, 0, len(dir))
+	for _, entry := range dir {
+		if entry.Type().IsRegular() {
+			digests = append(digests, ToDigest(entry.Name()))
+		}
+	}
+
+	return digests, nil
+}
+
+func (c *FileSystemCTF) SaveBlob(b blob.ReadOnlyBlob) (err error) {
+	digestable, ok := b.(blob.DigestAware)
+	if !ok {
+		return errors.New("blob does not have a digest that can be used to save it")
+	}
+
+	data, err := b.ReadCloser()
+	if err != nil {
+		return fmt.Errorf("unable to read blob: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, data.Close())
+	}()
+
+	dig, known := digestable.Digest()
+	if !known {
+		return errors.New("blob does not have a digest that can be used to save it")
+	}
+
+	return c.writeFile(filepath.Join(
+		BlobsDirectoryName,
+		ToBlobFileName(dig),
+	), data)
+}
+
+// ToBlobFileName converts a digest to a blob file name by replacing the ":" with ".", which is the
+// default separator for blobs in the CTF under BlobsDirectoryName.
+func ToBlobFileName(digest string) string {
```

Comment:
If this is something that the user could influence, I'm pretty sure we should be careful with replacing `.` otherwise we might expose ourselves to a directory traversal attack.

Reply:
im not sure how that can be used for exposure but im happy to be taught. I think this conversion is static and can never be modified

Reply:
What if the name is `.:/` or something. that would result in `..` involuntary providing a traverse once that's written to the filesystem.

Reply:
then that wouldnt be a valid digest anymore would it?

Reply:
Indeed. That's why I'm asking if this could be something a user would provide? Or if there is a valid digest check before calling this method or whether we should add one here? :)

Reply:
I think adding a valid digest is a good idea, ill add it

Reply:
added validation of digest to every call